### DISCUSSION TITLE
How to capture Rust stack trace using Sentry?

### DISCUSSION THREAD

**Discussion by AlienKevin** [fzyzcjy/flutter_rust_bridge#1487] [Q&A] [answered]
# Problem: Rust stack trace missing
Recently, I noticed quite a few `PanicException`s on my Sentry dashboard. I suspect this has to do with my refactoring which makes the initialization of the Rust api asynchronous. The stack trace captured by Sentry shows the Dart caller of the buggy api method but does not show where in the method the Panic occured in Rust:
```
basic.dart in FlutterRustBridgeBase._transformRust2DartMessage at line 135 within flutter_rust_bridge
In App
basic.dart in FlutterRustBridgeBase.executeNormal.<fn> at line 70 within flutter_rust_bridge
In App
Called from: <asynchronous suspension>
home_page.dart in _HomePageState.doSearch.<fn> at line 315 within wordshk
In App
```

Here's a screenshot of the PanicException on Sentry:
<img width="1333" alt="Screenshot 2023-12-23 at 7 00 43 PM" src="https://github.com/fzyzcjy/flutter_rust_bridge/assets/22850071/8acea7b9-5cd1-4f29-99ba-f2d3625be999">

# Current Sentry Setup
I'm using a standard Sentry setup which works well for capturing Dart stack traces:
```dart
main() async {
  WidgetsFlutterBinding.ensureInitialized();

  await SentryFlutter.init((options) {
    options.dsn = sentryDsn;
    options.tracesSampleRate = 0;
    options.addEventProcessor(CustomSentryEventProcessor());
  }, appRunner: runMyApp);
}

runMyApp({bool? firstTimeUser, Language? language}) async {
  try {
    ...
  } catch (exception, stackTrace) {
    await Sentry.captureException(
      exception,
      stackTrace: stackTrace,
    );
  }
}
```

# Question
**My setup apparently does not capture the Rust stack trace on any error originted in the Rust code**, as I checked that another FfiException error also has the same redacted stack trace. **How do I setup to capture full stack trace on the Rust side as well in Sentry?** I'm using V1 currently but I'm not sure if V2 will change the stack trace situation?

**Comment by fzyzcjy** [comment]
I personally did this (I also use sentry): https://cjycode.com/flutter_rust_bridge/guides/how-to/report-error
It is a v1-and-v2 feature, so it should be workable on v1 :)

EDIT: doc updated. In your case, maybe firstly try https://cjycode.com/flutter_rust_bridge/guides/how-to/stack-trace

![image](https://github.com/fzyzcjy/flutter_rust_bridge/assets/5236035/e24fd244-88fa-4413-95ab-711d35204c2f)

**Reply by AlienKevin** [reply]
I tried the enable stack trace suggestion from the doc but the Sentry trace is still redacted for some reason. Though I'm not sure if I understand the doc correctly. The doc says to turn on backtrace before initializing the bridge, I understand it as setting the env var before calling the Api constructor in my Lazy initializer.
> To pass Rust stack traces to flutter, you need to set RUST_BACKTRACE in the running application. For that simply add env::set_var("RUST_BACKTRACE", "1"); before initialising the bridge.

```rust
static API: Lazy<WordshkApi> =
    Lazy::new(|| { env::set_var("RUST_BACKTRACE", "1"); WordshkApi::new() });

pub fn init_api() -> () {
    let _ = &*API;
}

impl WordshkApi {
    fn new() -> WordshkApi { ... }
    ...
}
```

**Reply by fzyzcjy** [reply, marked as answer]
That looks like the correct way.

Oh I see, the stack trace is inside PanicException's message field. V1 does not provide toString (https://github.com/fzyzcjy/flutter_rust_bridge/blob/1a5b35a45b7ddca36abfa2294174fc26e1806b36/frb_dart/lib/src/basic.dart#L217) while v2 does. So in V2 you should see the details of the panic including stack, instead of a "Instance of PanciException" without any info.

**Reply by AlienKevin** [reply]
I see, thanks for the explanation. I will upgrade to V2.

**Reply by fzyzcjy** [reply]
You are welcome!

**Reply by AlienKevin** [reply]
I just upgraded to v2.0.0-dev.4 and tested the same code with the RUST_BACKTRACE=1. I printed the backtrace value to check if the env var is set correctly before creating an artificial panic. On the Sentry side, I still see the same redacted error message:
<img width="1332" alt="Screenshot 2023-12-25 at 12 07 13 PM" src="https://github.com/fzyzcjy/flutter_rust_bridge/assets/22850071/b5f7decb-c45e-4215-bcf8-173819853ce3">

**Reply by AlienKevin** [reply]
And as of v2.0.0-dev.4, `handler` is a `pub(crate)` so the example in the doc doesn't work because `handler::Error` is not exposed. `handler::ErrorHandler` seems to be replaced by `ErrorListener` but it is also hidden. Lastly, `ReportDartErrorHandler` seems to be gone in v2.0.0.

```rust
pub struct MyErrorHandler(ReportDartErrorHandler);

impl ErrorHandler for MyErrorHandler {
    fn handle_error(&self, port: i64, error: handler::Error) {
        send_error_to_your_backend(&error);
        self.0.handle_error(port, error)
    }

    ...
}
```

**Reply by fzyzcjy** [reply]
Hmm... Firstly, without sentry, is there stacktrace? i.e. if you print out PanicException directly to the screen?

Then, is stacktrace still there in release mode if you print it out?

> And as of v2.0.0-dev.4, handler is a pub(crate) so the example in the doc doesn't work because handler::Error is not exposed. handler::ErrorHandler seems to be replaced by ErrorListener but it is also hidden. Lastly, ReportDartErrorHandler seems to be gone in v2.0.0.

This is definitely a bug! (It did not have tests covered historically, so yes I forgot it... :/)
Feel free to create an issue!

**Reply by AlienKevin** [reply]
I'm seeing stack traces for some panics but not the artificial one I created. Probably something weird with how Sentry captures the panic message. I think the RUST_BACKTRACE=1 probably is enough for most things.

**Reply by fzyzcjy** [reply]
Hmm.... Btw, have you stripped your .so file? IIRC you are a open source project, and if do not care about increasing a bit of .apk size, you may try to avoid stripping your .so file. Then Sentry may be better at symbolicating.

> I'm seeing stack traces for some panics but not the artificial one I created.

Do you mean the one in PanicException.message, or the one captured by sentry? If the former, it is really weird... FRB does not do anything special to this imho

**Reply by AlienKevin** [reply]
> Hmm.... Btw, have you stripped your .so file?

I followed the default cargokit setup. Not sure if cargokit strips the .so file by default?

>Do you mean the one in PanicException.message, or the one captured by sentry?

I just tried again and found that the PanicException.message is displayed but the stack trace goes into internal code of FRB, rather than my Rust implementation.

<img width="1334" alt="Screenshot 2023-12-30 at 10 59 12 PM" src="https://github.com/fzyzcjy/flutter_rust_bridge/assets/22850071/3d9f56a7-07bf-493b-b7cb-ad8bfd178547">

The way I triggered a panic is quite simple. I basically added a panic to a search function that's called from Dart.
```rust
pub fn search(query: String) -> SearchResults {
    let api = API.lock().unwrap();
    match &mut search::search(...) {
        ...
        SearchResult::All(...) => {
            panic!("testing panic in search");
        }
    }
}
```

**Reply by fzyzcjy** [reply]
I will firstly add some tests to ensure it is not a bug of FRB -> https://github.com/fzyzcjy/flutter_rust_bridge/issues/1553

**Comment by MRam03** [comment]
To capture a Rust stack trace using Sentry, you can use the sentry crate, which provides integration with the Sentry error tracking service.

Add the sentry crate to your Cargo.toml file:

toml

[dependencies]
sentry = "0.32.1"
Initialize Sentry with your DSN (Data Source Name) in your Rust code:

rust

use sentry::integrations::panic::register_panic_handler;
use sentry::IntoDsn;

fn main() {
    let dsn = "https://your-sentry-dsn@sentry.io/your-project-id";
    sentry::init((dsn, sentry::ClientOptions::default()));

    // Optional: Set the release version
    sentry::configure_scope(|scope| {
        scope.set_release(env!("CARGO_PKG_VERSION"));
    });

    // Register the panic handler to capture panics
    register_panic_handler();
    
    // Your application logic goes here
}
Replace "https://your-sentry-dsn@sentry.io/your-project-id" with your actual Sentry DSN.

Use Sentry to capture errors and stack traces:

rust

use sentry::capture_message;

fn main() {
    // Your application logic goes here

    // Example of capturing an error with a custom message
    if let Err(err) = some_function() {
        capture_message(&err.to_string());
    }
}

fn some_function() -> Result<(), &'static str> {
    // Some code that might produce an error
    Err("An example error")
}
Replace some_function and the error message with your actual code.


[it course in Chennai](https://innovatetechnologies.co.in/), [it institute in Chennai](https://innovatetechnologies.co.in/), [it training center in Chennai](https://innovatetechnologies.co.in/)

**Comment by fzyzcjy** [comment]
@AlienKevin I think https://github.com/fzyzcjy/flutter_rust_bridge/discussions/1487#discussioncomment-8021471 looks like spam content, because the post has link to a training website, and the account is registered 1 hour ago, and the answer looks like generated.

What do you think? If you also think so, maybe we should delete it. (I hide it currently to avoid problems)

Screenshots

<details><summary>Details</summary>
<p>

![image](https://github.com/fzyzcjy/flutter_rust_bridge/assets/5236035/bbbbbba1-c39e-4f04-8ffb-086bd42fdbca)

click link gives:

![image](https://github.com/fzyzcjy/flutter_rust_bridge/assets/5236035/b7e0d93e-027a-4791-bf3f-73e8ef8e8d64)


![image](https://github.com/fzyzcjy/flutter_rust_bridge/assets/5236035/8b77b294-2edf-4efa-b2fd-137fc69a5f85)

</p>
</details>

**Reply by AlienKevin** [reply]
Yeah, I also think it's spam. Feel free to delete, thanks.

**Reply by fzyzcjy** [reply]
No worries, I already marked as spam (months before), and seems GitHub auto folds it.
